src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/string_view.hpp

```diff
@@ -14,92 +14,490 @@
+   public:
+    // Pointer to (non-const) character type
+    using pointer = Char*;
+    // Pointer to const-character type
+    using const_pointer = const Char*;
+    // Type representing the size of a string
+    using size_type = std::size_t;
+    // Type representing the offset within a string
+    using difference_type = std::ptrdiff_t;
+    // The type of the string character
+    using value_type = Char;
+
+    // Constant sentinel value to represent an impossible/invalid string position
+    static constexpr const size_type npos = static_cast<size_type>(-1);
+
+   private:
+    // Pointer to the beginning of the string being viewed
+    const_pointer _begin = nullptr;
+    // The size of the array that is being viewed via `_begin`
+    size_type _size = 0;
+    // Alias of our own type
+    using self_type = basic_string_view;
+
+    /**
+     * @brief If R is a type for which we want to permit implicit conversion,
+     * evaluates to the type `int`. Otherwise, is a substitution failure.
+     */
+    template <typename S>
+    using _enable_string_conversion = bsoncxx::detail::requires_t<
+        int,
+        // Don't eat our own copy/move constructor:
+        bsoncxx::detail::negation<bsoncxx::detail::is_alike<S, self_type>>,
+        // The range's value must be the same as our character type
+        std::is_same<typename bsoncxx::detail::remove_reference_t<S>::value_type, value_type>,
+        detail::is_string_like<S>>;
+
+   public:
+    using traits_type = Traits;
+    using reference = Char&;
+    using const_reference = const Char&;
+    using iterator = const_pointer;
+    using const_iterator = iterator;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    /**
+     * @brief Default constructor. Constructs to an empty/null string view
+     */
+    constexpr basic_string_view() = default;
+
+    /// Default copy/move/assign/destroy
+
+    /**
+     * @brief Construct a new string view from a pointer-to-character and an
+     * array length.
+     */
+    constexpr basic_string_view(const_pointer s, size_type count) noexcept
+        : _begin(s), _size(count) {}
+
+    /**
+     * @brief Construct a new string view from a C-style null-terminated character array.
+     *
+     * The string size is inferred as-if by strlen()
+     */
+    constexpr basic_string_view(const_pointer s) noexcept
+        : _begin(s), _size(traits_type::length(s)) {}
+
+    /**
+     * @brief Implicit conversion from string-like ranges.
+     *
+     * Requires that `StringLike` is a non-array contiguous range with the same
+     * value type as this string view, and is a std::string-like value.
+     */
+    template <typename StringLike, _enable_string_conversion<StringLike> = 0>
+    constexpr basic_string_view(StringLike&& str) noexcept
+        : _begin(str.data()), _size(str.size()) {}
```

Comment:
This inclusion is intentional to support compat with std::string+std::string_view

Reply:
Good point, I overlooked support for `std::string::operator std::string_view()`.

Rather than constraining the template to _any_ string-like type with SFINAE, a simple dedicated overload for `std::string` should work just fine:

```cpp
/**
 * @brief Implicit conversion from std::string ranges.
 */
constexpr basic_string_view(const std::basic_string<Char, Traits>& str) noexcept
    : _begin(str.data()), _size(str.size()) {}
```

I'm not sure supporting an implicit conversion from `std::string_view` to `bsoncxx::stdx::string_view` is desirable, given they are distinct implementations of the same vocabulary type. (Despite our efforts, they're still unlikely to be perfectly substitutable with one another.) We may want to encourage downstream code converting between the two types to be explicit with conversions so it's clear which `string_view` they're dealing with.

If we _do_ still want to support it, I think we should support conversion _both_ ways in C++17 mode:

```cpp
#if /* C++17 */

constexpr basic_string_view(const std::basic_string_view<Char, Traits>& str) noexcept
    : _begin(str.data()), _size(str.size()) {}

operator std::basic_string_view<Char, Traits>() const noexcept {
    return std::basic_string_view<Char, Traits>(data(), size());
}

#endif
```

This would allow the two to be fully interchangeable/convertable with each other:

```cpp
std::string_view std_sv;
bsoncxx::stdx::string_view bsoncxx_sv;

bsoncxx_sv = std_sv; // OK: ctor (currently handled by range ctor).
std_sv = bsoncxx_sv; // OK: conversion operator (currently not supported).
```

Thoughts?

Reply:
That may work.

In either case, supporting conversion to/from `std::string_view` may be problematic by ambiguating with the C++23 from-range converting constructors, since `bsoncxx::string_view` satisfies the constraints thereof. Time will tell.